numpy/core/src/multiarray/item_selection.c

```diff
@@ -2081,11 +2083,18 @@ PyArray_CountNonzero(PyArrayObject *self)
     NPY_BEGIN_THREADS_DEF;
 
     /* Special low-overhead version specific to the boolean type */
-    if (PyArray_DESCR(self)->type_num == NPY_BOOL) {
+    dtype = PyArray_DESCR(self);
+    if (dtype->type_num == NPY_BOOL) {
         return count_boolean_trues(PyArray_NDIM(self), PyArray_DATA(self),
                         PyArray_DIMS(self), PyArray_STRIDES(self));
     }
 
+    if (dtype->type_num == NPY_OBJECT || dtype->type_num == NPY_VOID ||
+            dtype->type_num >= NPY_NTYPES ||
+            PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI)) {
```

Comment:
After studying, you are right, I think I just need to check the flag. Object dtypes have `NPY_OBJECT_DTYPE_FLAGS` set which includes the PYAPI flag, and voids have it set too. For user-defined types it is up to the implementor.